Move all coffeescript to JS

naegelyd opened the pull request:
Fix #143.

Also, this updates the Cilantro version and updates Varify code to match changes in Cilantro.

naegelyd committed:
Port ui/controls to JS

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Port ui/modals to JS

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Port ui/tables to JS

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Remove last of the coffee script with minimal results ported

WARNING: This does not include a full port of the results workflow to JS. This is just meant to serve as a starting point so I can track my work in adding all the features to the results workflow again.

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add latest changes to result template from Cilantro

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Override the onRender method to show a custom result table

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Update to latest Cilantro build

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Update main to match Cilantro & register ResultDetails dialog

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add class name to dialog and remove nesting from styles

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add padding to panels to match body top offset

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add support for expanding/collapsing row based on cohort overflow

Also fixes an issue with the CSS which was being set on the wrong element after the conversion to JS.

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add files and references for new phenotypes dialog

NOTE: This contains only the most trivial implementation of the dialog itself. All that is there is a loading indicator. Actual content to follow.
Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Restore sample label display next to result count

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Move context sample label retrieval to utility method

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Show error & abort request if sample count isn't exactly 1

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add old phenotype code as a dialog registered in main

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Handle row clicks at the row level using tracked dialog

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Upgrade to Cilantro 2.2.21

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add custom export dialog with change columns button

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Upgrade to Cilantro 2.2.22

This fixes an issue where multi-type exports sometimes only exported the first type

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/models/phenotype.js:
```diff
@@ -6,9 +6,63 @@ define([
+
+        parse: function(attrs) {
+            var path, lastModified, phenotypeModified;
+            //Backbone.Model.prototype.parse.call(this, attrs);
```
You can remove this line since `Backbone.parse` is a no-op.

bruth commented on varify/static/scripts/javascript/src/models/phenotype.js:
```diff
@@ -6,9 +6,63 @@ define([
+        },
+
+        parsePriority: function(value) {
+            return parseInt(value.priority, 10) || this.lowestPriority + 1;
```
I realize this may be domain-specific, but why does this default to lowest plus one? Maybe a comment would help.

naegelyd replied:
Ha, I wish I knew the answer to this. I was just copying over code that @jeffmax wrote. @jeffmax can you drop a note here explaining the `+ 1` here and I will add the comment to the code.

jeffmax replied:
Sorry, I am just discovering I missed all these notifies. I think that +1 is a ill-conceived idea I had to make sure that values that were not integers came below everything else, but it's highly unlikely to happen and confusing, and even if it does happen, the lowest priority is fine. Will remove it.

bruth commented on varify/static/scripts/javascript/src/ui/controls.js:
```diff
@@ -1,8 +1,13 @@
+    'underscore',
+    './controls/sift',
+    './controls/polyphen'
+], function() {
```
`_` must be the first argument here to prevent use of a global.

naegelyd committed:
Remove unnecessary parse and fix errors in phenotype model

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add missing underscore argument in controls

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/ui/exporter.js:
```diff
@@ -0,0 +1,33 @@
+    var ExporterDialog = c.ui.ExporterDialog.extend({
+        template: 'varify/export/dialog',
+
+        _events: {
```
The `events` hash can be defined as a function since Backbone uses [`_.result`](http://underscorejs.org/#result) to resolve the object:

``` javascript
events: function() {
    return _.extend({ ... }, c.ui.ExporterDialog.prototype.events);
}
```

naegelyd committed:
Fix flake8 issues

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/ui/exporter.js:
```diff
@@ -0,0 +1,33 @@
+            c.ui.ExporterDialog.prototype.initialize.call(this);
+        },
+
+        onRender: function() {
```
This doesn't extend anything, so it shouldn't be necessary.

bruth commented on varify/static/scripts/javascript/src/ui/modals.js:
```diff
@@ -1,8 +1,13 @@
+    'underscore',
+    './modals/result',
+    './modals/phenotype'
+], function() {
```
Define `_` as an argument.

naegelyd committed:
Remove unnecessary onRender override in exporter

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Use a function to extend events

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/ui/modals/phenotype.js:
```diff
@@ -0,0 +1,291 @@
+
+            // We are about to attempt to reload the phenotypes so we want to
+            // start fresh and clear any old artifacts.
+            this.ui.warning.hide();
```
Would it be worth having a class applied to this view that hides all of these elements?

bruth commented on varify/static/scripts/javascript/src/ui/modals/result.js:
```diff
@@ -1,683 +1,859 @@
+    '../../models',
+    '../../utils',
+    '../../templates',
+    'cilantro/utils/numbers'
```
Using the minified Cilantro build, this will load all of Cilantro anyway. All the number utilities are accessible at `c.utils`.

naegelyd committed:
Add missing underscore argument in modals

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/ui/modals/result.js:
```diff
@@ -1,683 +1,859 @@
-      return valid;
-    };
+                        if (eff.hgvs_c) {
+                            content.push('' + eff.hgvs_c + ' ');
```
I've seen this in various places. I don't believe (but since since it's JS I wouldn't be surprised) a prefix string concatenation does anything different from a postfix one. Since you concat a space here, the former should not be needed.

naegelyd replied:
In this case, we could probably get away without the leading `''` but in some cases it is essential. For example, this example that I copied from my console. The postfix, in this case, just turns the leading result into a string while the prefix uses string concatenation throughout.

``` javascript
'' + 1 + 2 + 3
"123"
1 + 2 + 3 + ''
"6"
```

bruth replied:
Ah right. Most of this should all go away when moving to a template-centric approach. I wouldn't worry about this now.

naegelyd committed:
Fix errors and use selector for hiding elements on phenotype fetch

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented:
This is out of the scope of this PR, but there are many places where HTML is being concatenated and inserted. These views should be changed to layouts and the larger HTML strings change to their own views. This enables using templates (and therefore overriding them using the Cilantro template API by projects) and handles a lot of the `if (foo && foo.length)` handling since collection views handle all that already.

naegelyd commented:
Agree with HTML concatenation issue addressed above. In this PR, I just wanted to get this code ported over to get off of Coffeescript. Going forward, there are a lot of opportunities for improving the JS especially those you mention above.

bruth commented on varify/static/scripts/javascript/src/ui/tables.js:
```diff
@@ -1,8 +1,15 @@
+    './tables/header',
+    './tables/row',
+    './tables/table'
+], function() {
```
`_` missing.

bruth commented on varify/static/scripts/javascript/src/ui/tables/header.js:
```diff
@@ -1,107 +1,121 @@
-            } else {
-              f.unset('sort');
-              return f.unset('sort_index');
+            if ((concept != null) && !isNaN(concept)) {
```
Should be `_.isNaN`?

bruth replied:
Nevermind. Native function.. I'm a noob.

naegelyd committed:
Import cilantro as a whole instead of just utile

As noted by Byron, when using the minified Cilantro, the full version was being imported anyway so we might as well explicitly import the full version and reference Numbers at c.utils.Numbers

Signed-off-by: Don Naegely <naegelyd@gmail.com>

naegelyd committed:
Add missing underscore argument in tables

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/ui/workflows/results.js:
```diff
@@ -1,684 +1,101 @@
-        return this.$('.export-progress-container').hide();
-      }
-    };
+        _events: {
```
Same as above regarding using a function.

bruth commented on varify/static/scripts/javascript/src/ui/workflows/results.js:
```diff
@@ -1,684 +1,101 @@
-      }
-    };
+        onRender: function() {
+            $(document).on('scroll', this.onPageScroll);
```
This should not matter in practice since workflows are not closed.. but this handle should be unbound `onClose`.

naegelyd replied:
Won't this be unbound in [c.ui.ResultWorkflow.onClose](https://github.com/cbmi/cilantro/blob/master/src/js/cilantro/ui/workflows/results.js#L238-L240) since I am extending it?

bruth replied:
Yes, you are right. Forgot about that.

naegelyd committed:
Fix flake8 issues in bench/

Signed-off-by: Don Naegely <naegelyd@gmail.com>

bruth commented on varify/static/scripts/javascript/src/utils.js:
```diff
@@ -2,6 +2,54 @@
+         * end of the URL and be left with the root URL so we can use this to
+         * construct the result URLs later on.
+         */
+        return window.location.href.replace(new RegExp('/[^/]*/$'), '/');
```
Literals are preferred over use of the constructor: `/[^\/]*\/$/`, but I get the desire not to have to escape the forward slashes.

bruth commented on varify/static/scripts/javascript/src/utils.js:
```diff
@@ -168,13 +168,44 @@ define([], function() {
+         */
+        var samples = [], json;
+
+        if (context  && (json = context.get('json'))) {
```
Extra space after `context`.

naegelyd committed:
Use literal, add missing import, and remove extra space in utils

Signed-off-by: Don Naegely <naegelyd@gmail.com>
